File: src/jsMain/kotlin/com/scoreo/infrastructure/LocalStorageGameTypeRepository.kt
package com.scoreo.infrastructure

import com.scoreo.domain.model.GameType
import com.scoreo.domain.port.GameTypeRepository
import kotlinx.browser.localStorage
import kotlinx.serialization.encodeToString

private const val KEY = "scoreo_gametypes"

class LocalStorageGameTypeRepository : GameTypeRepository {
    override fun getAll(): List<GameType> =
        localStorage.getItem(KEY)
            ?.let { runCatching { scoreoJson.decodeFromString<List<GameType>>(it) }.getOrDefault(emptyList()) }
            ?: emptyList()

    override fun save(gameType: GameType) {
        val updated = getAll().toMutableList()
        val idx = updated.indexOfFirst { it.id == gameType.id }
        if (idx >= 0) updated[idx] = gameType else updated.add(gameType)
        localStorage.setItem(KEY, scoreoJson.encodeToString(updated))
    }

    override fun saveAll(gameTypes: List<GameType>) {
        val existing = getAll().toMutableList()
        for (gameType in gameTypes) {
            val idx = existing.indexOfFirst { it.id == gameType.id }
            if (idx >= 0) existing[idx] = gameType else existing.add(gameType)
        }
        localStorage.setItem(KEY, scoreoJson.encodeToString(existing))
    }

    override fun findById(id: String): GameType? = getAll().find { it.id == id }
}

File: src/infrastructure/localStorage/localStorageGameTypeRepository.ts
import type { GameType } from '../../domain/model/gameType'
import { GameTypeSchema } from '../../domain/model/gameType.schema'
import type { GameTypeRepository } from '../../domain/port/gameTypeRepository'

const KEY = 'scoreo_gametypes'

const GameTypesSchema = GameTypeSchema.array()

function readAll(): GameType[] {
  const raw = localStorage.getItem(KEY)
  if (raw === null) return []
  try {
    return GameTypesSchema.parse(JSON.parse(raw))
  } catch {
    return []
  }
}

function writeAll(gameTypes: GameType[]): void {
  localStorage.setItem(KEY, JSON.stringify(gameTypes))
}

export class LocalStorageGameTypeRepository implements GameTypeRepository {
  getAll(includeInactive = false): GameType[] {
    const all = readAll()
    return includeInactive ? all : all.filter((g) => g.active)
  }

  save(gameType: GameType): void {
    const updated = readAll()
    const idx = updated.findIndex((g) => g.id === gameType.id)
    if (idx >= 0) updated[idx] = gameType
    else updated.push(gameType)
    writeAll(updated)
  }

  saveAll(gameTypes: GameType[]): void {
    const existing = readAll()
    for (const gameType of gameTypes) {
      const idx = existing.findIndex((g) => g.id === gameType.id)
      if (idx >= 0) existing[idx] = gameType
      else existing.push(gameType)
    }
    writeAll(existing)
  }

  findById(id: string): GameType | undefined {
    return readAll().find((g) => g.id === id)
  }
}
